ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/endpoint/ServiceProviderEndpointVerification.kt

```diff
@@ -0,0 +1,149 @@
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.compliance.web.slo.endpoint
+
+import io.ktor.http.Parameters
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.RELAY_STATE
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SAML_REQUEST
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SAML_RESPONSE
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SIGNATURE
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SIG_ALG
+import org.apache.wss4j.common.saml.OpenSAMLUtil
+import org.codice.compliance.Common.Companion.getSingleLogoutLocation
+import org.codice.compliance.utils.TestCommon.Companion.DSA_SP_ENTITY_INFO
+import org.codice.compliance.utils.TestCommon.Companion.DSA_SP_ISSUER
+import org.codice.compliance.utils.TestCommon.Companion.RESPONDER
+import org.codice.compliance.utils.TestCommon.Companion.SAML_ENCODING
+import org.codice.compliance.utils.TestCommon.Companion.SUCCESS
+import org.codice.compliance.utils.TestCommon.Companion.samlObjectToString
+import org.codice.compliance.utils.sign.SimpleSign
+import org.codice.compliance.verification.binding.PostBindingVerifier
+import org.codice.compliance.verification.binding.PostFormVerifier
+import org.codice.compliance.verification.binding.RedirectBindingVerifier
+import org.codice.compliance.verification.core.requests.CoreLogoutRequestProtocolVerifier
+import org.codice.security.saml.SamlProtocol
+import org.codice.security.sign.Encoder
+import org.joda.time.DateTime
+import org.opensaml.saml.common.SAMLVersion
+import org.opensaml.saml.saml2.core.LogoutRequest
+import org.opensaml.saml.saml2.core.impl.IssuerBuilder
+import org.opensaml.saml.saml2.core.impl.LogoutResponseBuilder
+import org.opensaml.saml.saml2.core.impl.StatusBuilder
+import org.opensaml.saml.saml2.core.impl.StatusCodeBuilder
+import java.net.URLDecoder
+import java.nio.charset.StandardCharsets
+import java.util.UUID
+import javax.ws.rs.core.UriBuilder
+import kotlin.collections.set
+
+class ServiceProviderEndpointVerification(private val sendValidResponse: Boolean = true) {
+
+    private var relayState: String? = null
+    private val logoutResponse = LogoutResponseBuilder().buildObject().apply {
+        id = "a" + UUID.randomUUID().toString()
+        version = SAMLVersion.VERSION_20
+        issueInstant = DateTime()
+        issuer = IssuerBuilder().buildObject().apply { value = DSA_SP_ISSUER }
+        status = StatusBuilder().buildObject().apply {
+            StatusCodeBuilder().buildObject().apply {
+                value = if (sendValidResponse) SUCCESS else RESPONDER
+            }
+        }
+    }
+
+    /* Validations */
+
+    internal fun redirectRequestIsValid(queryParameters: Parameters): Boolean {
+        val request = queryParameters[SAML_REQUEST] ?: return false
+
+        val paramMap = mutableMapOf<String, String>()
```

Comment:
✏️ After looking into the `Parameters` type more, turns out it has a `.toMap()` method that converts it into a `Map<String, List<String>>`. From there we can just reduce that `List<String>` into a `String` type and voila!

```
    internal fun redirectRequestIsValid(queryParameters: Parameters): Boolean {
        queryParameters[SAML_REQUEST] ?: return false

        val paramMap = queryParameters.toMap().mapValues { (_, valueSet) ->
                valueSet.reduce { acc, value -> acc + value }
        }

        paramMap[RELAY_STATE]?.let {
            RedirectBindingVerifier().verifyLengthAndDecodeRelayState(it)
            relayState = it
        }

        val dom = RedirectBindingVerifier().decode(paramMap, false)

        ...
```